### PR TITLE
Wikidata and Wikipedia import

### PR DESCRIPTION
**Pull request by ringgaard**
Import Wikidata and Wikipedias from dump files provided by WikiMedia. The Wikidata dumps are bzip2 compressed json files and Wikipedia is bzip2-compressed XML files. Each item (and property) in Wikidata and each article in Wikipedia is converted into a SLING frame.

NB: The 5000-line calendar.sling file is just SLING meta-data about items for calendar dates (weekdays, months, days, years, decades, centuries, and millennia).

### CODE REVIEW

File: sling/nlp/wiki/wikipedia-importer.cc [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "sling/base/logging.h"
+#include "sling/base/types.h"
+#include "sling/frame/object.h"
+#include "sling/frame/store.h"
+#include "sling/nlp/wiki/wiki.h"
+#include "sling/nlp/wiki/wiki.h"
```
Comment (rahul1980) [inline sling/nlp/wiki/wikipedia-importer.cc]: duplicate

File: sling/nlp/wiki/wikipedia-importer.cc [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+#include "sling/stream/file-input.h"
+#include "sling/string/numbers.h"
+#include "sling/string/printf.h"
+#include "sling/task/frames.h"
+#include "sling/task/task.h"
+#include "sling/task/process.h"
```
Comment (rahul1980) [inline sling/nlp/wiki/wikipedia-importer.cc]: alphabetize

File: sling/nlp/wiki/wikipedia-importer.cc [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        builder.Add(n_redirect_pageid_, pageid);
+        builder.Add(n_redirect_title_, title);
+        if (!redirect_.empty()) {
+          builder.AddLink(n_redirect_link_, Wiki::Id(lang_, redirect_));
+        }
+
+        // Output frame on redirect channel.
+        if (redirects_channel_) {
+          Frame frame = builder.Create();
+          //LOG(INFO) << "Redirect: " << ToText(frame, 2);
```
Comment (rahul1980) [inline sling/nlp/wiki/wikipedia-importer.cc]: this line can be removed now

File: sling/nlp/kb/calendar.h [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  CalendarMap weekdays_;
+
+  // Months (1=January).
+  CalendarMap months_;
+
+  // Days of year where the key is month*100+day.
+  CalendarMap days_;
+
+  // Years where BCE are represented by negative numbers. There is no year 0 in
+  // the AD calendar although the is an item for this concept.
```
Comment (rahul1980) [inline sling/nlp/kb/calendar.h]: typo: the is -> there is

File: sling/nlp/kb/calendar.h [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  CalendarMap years_;
+
+  // Decades. The decades are numbered as year/10, e.g. the decade from
+  // 1970-1979 has 197 as the key. The last decade before AD, i.e. years 1-9 BC,
+  // has decade number -1. Likewise, the first decade after AD, i.e. years
+  // 1-9 AD, has number 0. These two decades only has nine years.
```
Comment (rahul1980) [inline sling/nlp/kb/calendar.h]: only has -> only have

File: sling/nlp/kb/calendar.cc [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+      }
+    }
+
+    case Date::MONTH: {
+      Text month = MonthName(date.month());
+      if (!month.empty()) {
+        if (!year.empty()) {
+          return StrCat(month, " ", year);
+        } else {
+          return StrCat(month, " ", date.year());
```
Comment (rahul1980) [inline sling/nlp/kb/calendar.cc]: should we not have a special case for date.year() < 0 ?
Similarly below

Comment (ringgaard) [reply sling/nlp/kb/calendar.cc]: We don't support relative dates, so we cannot just have "March 6" without a year.

File: sling/nlp/kb/calendar.h (hunk 4 above)
Comment (ringgaard) [reply sling/nlp/kb/calendar.h]: Fixed.

File: sling/nlp/kb/calendar.h (hunk 5 above)
Comment (ringgaard) [reply sling/nlp/kb/calendar.h]: Fixed.

File: sling/nlp/wiki/wiki.cc [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+#include <string>
+
+#include "sling/util/unicode.h"
+
+namespace sling {
+namespace nlp {
+
+// Language priority order.
+const char *Wiki::language_priority[] = {
+  "en", "da", "sv", "no", "de", "fr", "es", "it", "nl", "pt", "pl", "fi",
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/wiki/wiki.cc]: I have chosen the priority based on a mix of the size of the Wikipedia for the language and my ability to understand the language, and minusta on todella vaikeaa lukea suomalaista.

File: sling/nlp/wiki/wikipedia-importer.cc (hunk 1 above)
Comment (ringgaard) [reply sling/nlp/wiki/wikipedia-importer.cc]: Fixed.

File: sling/nlp/wiki/wikipedia-importer.cc (hunk 2 above)
Comment (ringgaard) [reply sling/nlp/wiki/wikipedia-importer.cc]: Done.

File: sling/nlp/wiki/wikipedia-importer.cc (hunk 3 above)
Comment (ringgaard) [reply sling/nlp/wiki/wikipedia-importer.cc]: Removed.

File: sling/web/xml-parser.cc [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+                  // End of comment found, output comment.
+                  bufptr_ -= 3;
+                  AddText(0);
+                  if (!Comment(txtptr_)) return false;
+                  bufptr_ = txtptr_;
+                  txtptr_ = nullptr;
+                  ch = ReadChar();
+                  break;
+                }
+                else {
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/web/xml-parser.cc]: Fixed.

File: sling/web/xml-parser.h [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  const char *Get(const char *name, const char *defval) const;
+  const char *Get(const char *name) const;
+  void Clear() {
+    name = nullptr;
+    attrs.clear();
+  }
+};
+
+// XML event-driven parser. The XML input is parsed and the callbacks are called
+// when incrementally as each constituent of the XML file has been parsed. To
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/web/xml-parser.h]: I have rephrased this comment to be a little more clear.

File: sling/web/xml-parser.h [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+// use the parser, a subclass of the XMLParser is implemented implemeting the
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/web/xml-parser.h]: see above

**Pull request merged**